### PR TITLE
Refreshing deconstruction design

### PR DESCRIPTION
**Pull request by jcouv**
The design doc is out-of-date (new grammar was introduced, d-assignments return a value, tuple literals in deconstruction have their type inferred, etc). 

@dotnet/roslyn-compiler  for review. This is a doc-only change.

### CODE REVIEW

File: docs/features/deconstruction.md [hunk 1]
```diff
@@ -65,15 +64,17 @@ class C
 
 ###Deconstruction-assignment (deconstruction into existing variables):
 
-This doesn't introduce any changes to the language grammar. We have an `assignment-expression` (also simply called `assignment` in the C# grammar) where the `unary-expression` (the left-hand-side) is a `tuple-literal`.
-In short, what this does in the general case is find a `Deconstruct` method on the expression on the right-hand-side of the assignment, invoke it, collect its `out` parameters and assign them to the variables on the left-hand-side. And in the special case where the expression on the right-hand-side is a tuple (tuple literal or tuple type), then the elements of the tuple can be assigned to the variables on the left-hand-side without needing to call `Deconstruct`.
+This doesn't introduce any changes to the language grammar. We have an `assignment-expression` (also simply called `assignment` in the C# grammar) where the `unary-expression` (the left-hand-side) is a `tuple-expression`.
```
Comment (gafter) [inline docs/features/deconstruction.md]: >` [](start = 71, length = 1)

Use backticks to get a code font, typically when showing verbatim code. Here you probably want italics, so use an asterisk to surround the names of syntactic productions.

File: docs/features/deconstruction.md [hunk 2]
```diff
@@ -129,65 +130,47 @@ evaluate side-effect on the left-hand-side variables
...
+###Deconstruction-declaration (deconstruction into new variables):
+
+Deconstruction-declarations can be thought of as two steps: (1) declaring new locals, and (2) applying a deconstruction-assignment into those locals.
+The syntax for deconstruction is more involved and multiple forms are allowed. The simplest case is `(int x, string y)`. Variants include nested declarations like `(int x, (string y, long z))` (which declares 3 locals) and implicitly typed declarations like `(var x, var y)`. The latter can also be written using the shorthand `var (x, y)`.
 
+As in the case of deconstruction-declarations, tuple expressions on the right-hand-side have their type inferred from the left-hand-side. With deconstruction-declaration, this is also the case, except that any type is `var` in the left-hand-side, then the natural type of the element in the right-hand-side is used.
+For example, in `(string x, byte y, var z) = (null, 1, 2);`, `null` has type `string`, the literal `1` has type `byte` (inferred from `y`) and the literal `2` has type `int` (its natural type).
 
-###Deconstruction-declaration (deconstruction into new variables):
+###Grammar changes
```
Comment (gafter) [inline docs/features/deconstruction.md]: You should probably mention, somewhere, that when the syntax `type ( variable_designation, ... )` is used, *type* must be `var`.

**Comment by gafter** [comment]
:shipit:

**Comment by jcouv** [comment]
Thanks. I updated the doc with both suggestions. I'll merge once green.

**Pull request merged**